selfhost/parser.jakt

```diff
@@ -155,6 +170,164 @@ boxed enum ParsedStatement {
     Try(stmt: ParsedStatement, error_name: String, error_span: Span, catch_block: ParsedBlock, span: Span)
     Garbage(Span)
 
+    function equals(this, anon parsed_statement: ParsedStatement) -> bool => match this {
+        Expression(expr) => {
+            let lhs_expr = expr
+            yield match parsed_statement {
+                Expression(expr) => lhs_expr.equals(expr)
+                else => false
+            }
+        }
+        Defer(statement) => {
+            let lhs_statement = statement
+            yield match parsed_statement {
+                Defer(statement) => lhs_statement.equals(statement)
+                else => false
+            }
+        }
+        UnsafeBlock(block) => {
+            let lhs_block = block
+            yield match parsed_statement {
+                UnsafeBlock(block) => lhs_block.equals(block)
+                else => false
+            }
+        }
+        VarDecl(var, init) => {
+            let lhs_var = var
+            let lhs_init = init
+            yield match parsed_statement {
+                VarDecl(var, init) => lhs_var.equals(var) and lhs_init.equals(init)
+                else => false
+            }
+        }
+        If(condition, then_block, else_statement) => {
+            let lhs_condition = condition
+            let lhs_then_block = then_block
+            let lhs_else_statement = else_statement
+            yield match parsed_statement {
+                If(condition, then_block, else_statement) => {
+                    if not (lhs_condition.equals(condition) and lhs_then_block.equals(then_block)) {
+                        return false
+                    }
+                    if not lhs_else_statement.has_value() {
+                        return not else_statement.has_value()
+                    } else {
+                        if not else_statement.has_value() {
+                            return false
+                        }
+                        if lhs_else_statement!.equals(else_statement!) {
+                            return true
+                        }
+                        return false
+                    }
+                }
+                else => false
+            }
+        }
+        Block(block) => {
+            let lhs_block = block
+            yield match parsed_statement {
+                Block(block) => lhs_block.equals(block)
+                else => false
+            }
+        }
+        Loop(block) => {
+            let lhs_block = block
+            yield match parsed_statement {
+                Loop(block) => lhs_block.equals(block)
+                else => false
+            }
+        }
+        While(condition, block) => {
+            let lhs_condition = condition
+            let lhs_block = block
+            yield match parsed_statement {
+                While(condition, block) => lhs_condition.equals(condition) and lhs_block.equals(block)
+                else => false
+            }
+        }
```

Comment:
You can actually write these a bit more compactly:
```suggestion
        While(condition: lhs_condition, block: lhs_block) => {
            yield match parsed_statement {
                While(condition, block) => lhs_condition.equals(condition) and lhs_block.equals(block)
                else => false
            }
        }
```

Reply:
Changed into the compact form as suggested